---
title: "<em>FibroDB</em> Documentation"
subtitle: "v1.0.0"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: "cosmo"
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false 
    toc_depth: 2
    self_contained: false
params:
  object: "placeholder"
---

```{r setup, echo=FALSE, include=FALSE}
library(tidyverse)
library(shiny)
source("../utils.R")
```

![](logo.png){align="right" style="border: none; float: right;" width="360"}

**FibroDB** is a web database for **accessing and exploring fibroblast expression data.**. The data for this database was derived from four studies profiled by *Ilieva et al., 2021*. **FibroDB** is the work of the [Uchida laboratory](https://heartlncrna.github.io/){target="_blank"}, Center for RNA Medicine, Aalborg University, and the [Bioinformatics Research Network](http://bio-net.dev/){target="_blank"}.

Key features:

1. View transcriptomics data across four fibroblast studies

2. Explore differential gene expression results

3. Download standardized and reprocessed datasets

## Datasets

The current version of **FibroDB** contains the following datasets:

- [GSE97829](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE97829){target="_blank"}: RNA-seq data of the MRC5 lung fibroblastic cell line treated with TGF-β.

- [GSE123018](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123018){target="_blank"}: Time-course RNA-seq data of cardiac fibroblasts with TGF-β.

- [GSE149413](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149413){target="_blank"}: RNA-seq data from fibroblasts isolated from pairs of chronic thromboembolic pulmonary hypertension (CTEPH) thrombus and pulmonary artery adventitia of the same CTEPH patient (four CTEPH patients in total.

- [GSE140523](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140523){target="_blank"}: RNA-seq data set on fibroblasts from abdomen, upper gingiva, lung, soft palate, scalp, trachea, and vocal fold.


## FibroDB Interfaces

![**The 'Results' page interface**. (**A**) Controls the study which is displayed. (**B**) The Results table which displays the results of differential gene expression analysis for each gene in the study. Selecting a row will change the output in (**D**). (**C**) Selector which controls the type of expression data displayed in **D**. (**D**) Interactive box plots showing the distribution of gene expression values across samples within each biological condition in the selected study.](usage_fibrodb_1.small.png){align="right" style="border: none; float: none;" width="1000"}

![**DGE volcano plots**. This interface displays the results of DGE analysis as a volcano plot (highlighted). Selecting a row in the 'Results Table' will cause the corresponding gene to be highlighted in the volcano plot.](usage_fibrodb_2.small.png){align="right" style="border: none; float: none;" width="1000"}

![**DGE Heatmap**. This interface displays the top hits from DGE analysis as a heatmap for the selected study.](usage_fibrodb_3_small.png){align="right" style="border: none; float: none;" width="1000"}

![**Pathway enrichment plot**. This interface displays the top results from KEGG pathway analysis as a heatmap. A drop down is available for controlling the displayed enrichment metric. "Combined.Score" - a combination of the p value from Fisher's exact test and z score from expected enrichment; "Odds.Ratio" - odds ratio from Fisher's exist test; "Padj (-log10)" - Adjusted p value from enrichment analysis, log transformed. For more info, see the [enrichr](https://maayanlab.cloud/Enrichr/) documentation.](usage_fibrodb_4_small.png){align="right" style="border: none; float: none;" width="1000"}

![**DGE Venn Diagrams**. This interface displays Venn diagrams which are used to compare the results of DGE analysis from all studies. A drop down menu is used to control whether Over- or Under-expressed DEGs are compared.](usage_fibrodb_5_small.png){align="right" style="border: none; float: none;" width="1000"}

## Terminology

- *DGE*: Differential Gene Expression
  - Differential Gene Expression refers to a significant difference in the expression of a gene between two conditions of interest. In this study, we calculated DGE using the [egdeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html){target="_blank"} R/Bioconductor package.
- *RPKM*: Reads Per Kilobase of transcript, per Million mapped reads
  - Read counts normalized based on transcript length and read depth. [More info](https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/){target="_blank"}.
- *CPM*: Counts per Million
  - Read counts normalized by library size. In this case, they were normalized using the moderated counts-per-million method in [egdeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html){target="_blank"}.
- *TPM*: Transcripts per Million
  - Read counts normalized based on transcript length and read depth (different from RPKM). [More info](https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/){target="_blank"}.


## Bugs

FibroDB is a new database, which means bugs may occasionally occur. If you encounter any bugs or unexpected behavior, please [open an issue](https://github.com/Bishop-Laboratory/FibroDB/issues/) on the RLBase GitHub repo and describe, in as much detail as possible, the following:

1.  What you expected FibroDB to do.
2.  What FibroDB did and why it was unexpected.
3.  Any error messages you received (along with screenshots).

## License and attribution

FibroDB is licensed under an MIT license and we ask that you please cite FibroDB in any published work like so:

> Ilieva et al., FibroDB: Expression Analysis of Protein-Coding and Long Non-Coding RNA Genes in Fibrosis, 2022
