<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2022-01-12" />

<title>FibroDB Documentation</title>

<script src="documentation_files/header-attrs-2.11/header-attrs.js"></script>
<script src="documentation_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="documentation_files/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="documentation_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="documentation_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="documentation_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="documentation_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="documentation_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="documentation_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="documentation_files/navigation-1.1/tabsets.js"></script>
<link href="documentation_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="documentation_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore"><em>FibroDB</em> Documentation</h1>
<h3 class="subtitle">v1.0.0</h3>
<h4 class="date">2022-01-12</h4>

</div>


<p><img src="logo.png" align="right" style="border: none; float: right;" width="360" /></p>
<p><strong>FibroDB</strong> is a web database for <strong>accessing and exploring fibroblast expression data.</strong>. The data for this database was derived from four studies profiled by <em>Ilieva et al., 2021</em>. <strong>FibroDB</strong> is the work of the <a href="https://heartlncrna.github.io/" target="_blank">Uchida laboratory</a>, Center for RNA Medicine, Aalborg University, and the <a href="http://bio-net.dev/" target="_blank">Bioinformatics Research Network</a>.</p>
<p>Key features:</p>
<ol style="list-style-type: decimal">
<li><p>View transcriptomics data across four fibroblast studies</p></li>
<li><p>Explore differential gene expression results</p></li>
<li><p>Download standardized and reprocessed datasets</p></li>
</ol>
<div id="datasets" class="section level2">
<h2>Datasets</h2>
<p>The current version of <strong>FibroDB</strong> contains the following datasets:</p>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE97829" target="_blank">GSE97829</a>: RNA-seq data of the MRC5 lung fibroblastic cell line treated with TGF-β.</p>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123018" target="_blank">GSE123018</a>: Time-course RNA-seq data of cardiac fibroblasts with TGF-β.</p>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149413" target="_blank">GSE149413</a>: RNA-seq data from fibroblasts isolated from pairs of chronic thromboembolic pulmonary hypertension (CTEPH) thrombus and pulmonary artery adventitia of the same CTEPH patient (four CTEPH patients in total.</p>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140523" target="_blank">GSE140523</a>: RNA-seq data set on fibroblasts from abdomen, upper gingiva, lung, soft palate, scalp, trachea, and vocal fold.</p>
</div>
<div id="fibrodb-interfaces" class="section level2">
<h2>FibroDB Interfaces</h2>
<div class="figure">
<img src="usage_fibrodb_1.small.png" align="right" style="border: none; float: none;" width="1000" alt="" />
<p class="caption"><strong>The ‘Results’ page interface</strong>. (<strong>A</strong>) Controls the study which is displayed. (<strong>B</strong>) The Results table which displays the results of differential gene expression analysis for each gene in the study. Selecting a row will change the output in (<strong>D</strong>). (<strong>C</strong>) Selector which controls the type of expression data displayed in <strong>D</strong>. (<strong>D</strong>) Interactive box plots showing the distribution of gene expression values across samples within each biological condition in the selected study.</p>
</div>
<div class="figure">
<img src="usage_fibrodb_2.small.png" align="right" style="border: none; float: none;" width="1000" alt="" />
<p class="caption"><strong>DGE volcano plots</strong>. This interface displays the results of DGE analysis as a volcano plot (highlighted). Selecting a row in the ‘Results Table’ will cause the corresponding gene to be highlighted in the volcano plot.</p>
</div>
<div class="figure">
<img src="usage_fibrodb_3_small.png" align="right" style="border: none; float: none;" width="1000" alt="" />
<p class="caption"><strong>DGE Heatmap</strong>. This interface displays the top hits from DGE analysis as a heatmap for the selected study.</p>
</div>
<div class="figure">
<img src="usage_fibrodb_4_small.png" align="right" style="border: none; float: none;" width="1000" alt="" />
<p class="caption"><strong>Pathway enrichment plot</strong>. This interface displays the top results from KEGG pathway analysis as a heatmap. A drop down is available for controlling the displayed enrichment metric. “Combined.Score” - a combination of the p value from Fisher’s exact test and z score from expected enrichment; “Odds.Ratio” - odds ratio from Fisher’s exist test; “Padj (-log10)” - Adjusted p value from enrichment analysis, log transformed. For more info, see the <a href="https://maayanlab.cloud/Enrichr/">enrichr</a> documentation.</p>
</div>
<div class="figure">
<img src="usage_fibrodb_5_small.png" align="right" style="border: none; float: none;" width="1000" alt="" />
<p class="caption"><strong>DGE Venn Diagrams</strong>. This interface displays Venn diagrams which are used to compare the results of DGE analysis from all studies. A drop down menu is used to control whether Over- or Under-expressed DEGs are compared.</p>
</div>
</div>
<div id="terminology" class="section level2">
<h2>Terminology</h2>
<ul>
<li><em>DGE</em>: Differential Gene Expression
<ul>
<li>Differential Gene Expression refers to a significant difference in the expression of a gene between two conditions of interest. In this study, we calculated DGE using the <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html" target="_blank">egdeR</a> R/Bioconductor package.</li>
</ul></li>
<li><em>RPKM</em>: Reads Per Kilobase of transcript, per Million mapped reads
<ul>
<li>Read counts normalized based on transcript length and read depth. <a href="https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/" target="_blank">More info</a>.</li>
</ul></li>
<li><em>CPM</em>: Counts per Million
<ul>
<li>Read counts normalized by library size. In this case, they were normalized using the moderated counts-per-million method in <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html" target="_blank">egdeR</a>.</li>
</ul></li>
<li><em>TPM</em>: Transcripts per Million
<ul>
<li>Read counts normalized based on transcript length and read depth (different from RPKM). <a href="https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/" target="_blank">More info</a>.</li>
</ul></li>
</ul>
</div>
<div id="bugs" class="section level2">
<h2>Bugs</h2>
<p>FibroDB is a new database, which means bugs may occasionally occur. If you encounter any bugs or unexpected behavior, please <a href="https://github.com/Bishop-Laboratory/FibroDB/issues/">open an issue</a> on the RLBase GitHub repo and describe, in as much detail as possible, the following:</p>
<ol style="list-style-type: decimal">
<li>What you expected FibroDB to do.</li>
<li>What FibroDB did and why it was unexpected.</li>
<li>Any error messages you received (along with screenshots).</li>
</ol>
</div>
<div id="license-and-attribution" class="section level2">
<h2>License and attribution</h2>
<p>FibroDB is licensed under an MIT license and we ask that you please cite FibroDB in any published work like so:</p>
<blockquote>
<p>Ilieva et al., FibroDB: Expression Analysis of Protein-Coding and Long Non-Coding RNA Genes in Fibrosis, 2022</p>
</blockquote>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
